src/bdsp/bdsp.c

```diff
@@ -0,0 +1,98 @@
+/*
+ * Pokémon Brilliant Diamond and Shining Pearl automation
+ */
+
+#include <util/delay.h>
+
+#include "automation-utils.h"
+#include "user-io.h"
+
+/* Static functions */
+static void temporary_control(void);
+static void shiny_shaymin(void);
+
+
+int main(void)
+{
+	init_automation();
+	init_led_button();
+
+	/* Initial beep to confirm that the buzzer works */
+	beep();
+
+	/* Wait for the user to press the button (should be on the Switch main menu) */
+	count_button_presses(100, 100);
+
+	/* Set the virtual controller as controller 1 */
+	switch_controller(REAL_TO_VIRT);
+
+	for (;;) {
+		/* Set the LEDs, and make sure automation is paused while in the
+		   menu */
+		set_leds(BOTH_LEDS);
+		pause_automation();
+
+		/* Feature selection menu */
+		uint8_t count = count_button_presses(100, 900);
+
+		for (uint8_t i = 0 ; i < count ; i += 1) {
+			beep();
+			_delay_ms(100);
+		}
+
+		switch (count) {
+			case 1:
+				temporary_control();
+				break;
+			case 2:
+				shiny_shaymin();
+				break;
+
+
+			default:
+				/* Wrong selection */
+				delay(100, 200, 1500);
+				break;
+		}
+	}
+}
+
+/*
+ * Temporary gives back control to the user by performing controller switch.
+ */
+void temporary_control(void)
+{
+	set_leds(NO_LEDS);
+
+	/* Allow the user to connect their controller back as controller 1 */
+	switch_controller(VIRT_TO_REAL);
+
+	/* Wait for the user to press the button (should be on the Switch main menu) */
+	count_button_presses(100, 100);
+
+	/* Set the virtual controller as controller 1 */
+	switch_controller(REAL_TO_VIRT);
+}
+
+/*
+ * Macro for hunting shiny shaymin
+ */
+void shiny_shaymin(void)
+{
+	for (;;) {
+		SEND_BUTTON_SEQUENCE(
+			{ BT_NONE, DP_TOP, SEQ_HOLD, 2 }, /* Navigate to "run" */
+			{ BT_A, DP_NEUTRAL, SEQ_HOLD, 2 }, /* Select "run" */
+			{ BT_NONE, DP_NEUTRAL, SEQ_HOLD, 132 }, /* Wait for scene change */
+			{ BT_A, DP_NEUTRAL, SEQ_HOLD, 8 }, /* Close "Shaymin disappeared dialog "*/
+			{ BT_P, DP_NEUTRAL, SEQ_HOLD, 8 }, /* Get on bike */
+			{ BT_NONE, DP_BOTTOM, SEQ_HOLD, 64 }, /* Ride down out of area */
+			{ BT_NONE, DP_TOP, SEQ_HOLD, 65 }, /* Ride back to Shaymin */
+			{ BT_A, DP_NEUTRAL, SEQ_HOLD, 2 }, /* Talk to Shaymin */
+			{ BT_NONE, DP_NEUTRAL, SEQ_HOLD, 64 }, /* Wait for Shaymin cry to finish */
+			{ BT_A, DP_NEUTRAL, SEQ_HOLD, 2 }, /* Start Shaymin encounter */
+			{ BT_NONE, DP_NEUTRAL, SEQ_HOLD, 512 } /* Wait for menu to pop up again */
+		);
+	}
```

Comment:
Instead of an infinite loop requiring a reset, I think it would be better to wait a few second after each sequence to allow the user to press the button. When it’s pressed, give control to the user, and then (when the user gives control back) break out of the loop to return to the main menu. See https://github.com/VinDuv/switch-arduino-controller/blob/master/src/swsh/swsh.c#L289 for an example.

Reply:
I think this will work well. I may do something with the lights at this time so it's obvious to the user that *this* is the time they can stop the macro, since this macro in particular (the Shaymin one) takes a fair bit of time to execute. Subsequent ones, like Arceus, will take even longer.

Reply:
Thinking about this, maybe it would be nice to have some sort of “emergency interrupt” function that would check on every cycle if the button is pressed, and if it is, interrupt the automation and break out of the current automation loop (using `longjmp`). That way, the user could press the button at any point to immediately interrupt the macro. I may try to add that if I have the time.

Reply:
That sounds great. I would definitely utilize that if it existed.

Reply:
I updated the pull request to only include the single commit which changed the Makefile. I am keeping my bdsp additions on their own branch in my fork for now. I'm not sure when I'll get around to making the reset change. It probably won't be until I implement a Darkrai or Arceus hunting macro. When I do I'll make a PR.